overlord/state/taskrunner.go

```diff
@@ -63,29 +63,45 @@ func (r *TaskRunner) Handlers() map[string]HandlerFunc {
 	return r.handlers
 }
 
+// taskFail marks task t and all tasks waiting (directly and indirectly on it) as in ErrorStatus.
+func taskFail(t *Task) {
+	mark := make([]*Task, 1, 10)
+	mark[0] = t
+	i := 0
+	for i < len(mark) {
+		mark[i].SetStatus(ErrorStatus)
+		mark = append(mark, mark[i].HaltTasks()...)
+		i++
+	}
+}
+
 // run must be called with the state lock in place
 func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 	task.SetStatus(RunningStatus) // could have been set to waiting
-	tomb := &tomb.Tomb{}
-	r.tombs[task.ID()] = tomb
-	tomb.Go(func() error {
-		err := fn(task, tomb)
+	tb := &tomb.Tomb{}
+	r.tombs[task.ID()] = tb
+	tb.Go(func() error {
+		err := fn(task, tb)
 
 		r.state.Lock()
 		defer r.state.Unlock()
-		if err == nil {
+		halted := task.HaltTasks()
+		if err == nil && tb.Err() == tomb.ErrStillAlive {
```

Comment:
There are three cases which are being either mishandled by this logic or that make the implementation difficult to understand:

* The handler was asked to stop but succeeded doing its intended goal despite it
* The handler was asked to stop but would like to try again
* The handler would just like to try again later anyway, whether it was asked to stop or not.

We probably want something around these lines:
```
tb.Kill(fn(task, tb))
...
switch tb.Err() {
case Retry:
        // Do nothing. Handler asked to try again later.
case nil:
        task.SetStatus(DoneStatus)
        if len(task.HaltTasks()) > 0 {
                r.state.EnsureBefore(0)
        }
default:
        taskFail(task)
}
```